### DISCUSSION TITLE
skip xAxis ticks for no data

### DISCUSSION THREAD

**Discussion by luissimoes** [dbuezas/lovelace-plotly-graph-card#365] [Q&A] [answered]
I am collecting daily data for 4 sensors and showing it in a bar chart.
What I want is to show the data only once ... eg midnight or any time actually since this is a delta.

What I have is in the figure... i.e. the x axis shows ticks at times between the data.
- nticks doesn't solve my problem because the *gaps* between the bar groups stays the same... (looks very bad)
- range breaks: enabled: true doesn't solve it either....

in Apex I had the following function for the xAxis (as formatter)
`      formatter: |
        EVAL:function(value) {
          result = new Date(value).toLocaleDateString("us",{month:"short", day:"2-digit"});
          return result
        } `

<img width="957" alt="Screenshot 2024-01-26 at 15 58 39" src="https://github.com/dbuezas/lovelace-plotly-graph-card/assets/42578174/dfb1adc2-887c-4450-a763-dc9205d37d0a">

____ Add-On

I added "resample: 24h" to the config, and got a "nearer" result..... (see below) but now the current value for today do not show...
:-(

<img width="957" alt="Screenshot 2024-01-26 at 17 58 18" src="https://github.com/dbuezas/lovelace-plotly-graph-card/assets/42578174/4bff832b-9e70-4431-bdd5-7c66a454fda8">

please any hint?
many thanks!

Luis

**Comment by dbuezas** [comment]
I think you can use a $fn or $ex to set`layout: xaxis: tickvals` based on your data.
Alternatively you could shift the data 23 hours forwards, then resample, and then offset it back to fix your second approach.
Let me know if you succeed!

**Comment by luissimoes** [comment]
Hi
:-(.... no luck... I have now looked at this for hours and hours - what just proves I am a complete ignorant in the matter....
I tried 
- tickmode: linear, dtick and tick0 set with a javascript to be 5 days before "now"
- tickmode: array.... 

no luck....

I am doing something wrong for sure.... but don't now what

Also
-not sure what your second suggestion is... because the 23 hours forward is not done not at filter level right?
How do you mean shift the data forwards *and* then back?

thanks!

**Reply by dbuezas** [reply]
Don't give up! :)

For the shifting hack:
```yaml
- map_x:+ x+23*60*60*1000
- resample: 1d
- map_x: +x-23*60*60*1000
```
For tickvals see https://plotly.com/javascript/tick-formatting/#tickmode---array

The idea is to generate them via $ex

**Reply by luissimoes** [reply]
Hi :-)
I don't give up, as I think Plotly is great. But it's proving difficult for this case - always when there is a "statistic" involved I don't see any data in the last day - see figure 1, where I implemented the suggestion with the time shift.



.... may be you *have a suggestion* as my case is (I think) pretty standard...?

=> My case
- I have sensors: energy, increasing, that reset at 00:00.
- if I represent them just like that I get the following graph (figure 2) which is ok... but I want
1) **just the max value,** which is of course at midnight - (end of the day)
2) **represent that value by a bar**
3) the current day, shows then the value as it is at the time.... 

I think this is very easy... but I how can I get the last (max) value and show it as a bar?

many thanks for any support!

Luis

figure 1 (with suggestion of time shifting and resample)
<img width="810" alt="Screenshot 2024-01-28 at 00 06 35" src="https://github.com/dbuezas/lovelace-plotly-graph-card/assets/42578174/db690464-944f-4b3b-87fe-086e0f251f12">


figur3 2 (sensors without statistics but not in bar ....  see my comments/wishes)
<img width="482" alt="Screenshot 2024-01-27 at 16 03 11" src="https://github.com/dbuezas/lovelace-plotly-graph-card/assets/42578174/7f7b6504-d06d-4e3a-8b34-7e428e5893dd">

**Reply by dbuezas** [reply]
Oh, I just realized that the original issue you have has a different reason.

Try this:
```yaml
filters:
  - filter: i>0
```

This will remove the first datapoint which has the wrong date (home assistant returns a fake datapoint at the beginning of the requested range)

With that, the widths of the bars and the xaxis tick text should be fixed.

Instead of resampling, you can use statistics with `period: day`, and to add the current value of the day you can append an extra datapoint with the right time stamp: see here https://github.com/dbuezas/lovelace-plotly-graph-card/discussions/299#discussioncomment-6530739

**Reply by luissimoes** [reply, marked as answer]
NOW we are there! **thanks to YOU**!  :-)

Actually I didn't need the "i >0 " just the part in the filter to take the last point out (i < xs.length - 1)

What I did is to use 
statistic: sum
period: day
 (I had statistic: state --- and this cause the extra x marks to be shown, even with the i > 0)

Many thanks for your wonderful support and help!

The only thing now is the indicators.... 
I was using at the layout level  "fn: $fn ({ys, vars}) => {vars.fromnetz = ys.at(-1);}" to get the value.
but now... how can I use it ?

I tried the following (added cards to the input list and see the assignment below....); but this doest bring any value.
```yaml
      - fn: |
          ({ys,xs,hass, vars}) => ({
            xs: [...xs, new Date().setHours(0,0,0)],
            ys: [...ys, hass.states['sensor.from_netz_daily'].state],
            vars: vars.fromnetz = ys.at(-1),
          })
```
I thought about creating a new entity for the same sensor, which doesn't display, and I just use it to set the indicator.... but perhaps there is a more elegant solution using the existing fb filter block

Obrigado!!!

**Reply by dbuezas** [reply]
Happy to hear!
For the indicator, see: https://github.com/dbuezas/lovelace-plotly-graph-card/discussions/364#discussioncomment-8231838

**Reply by luissimoes** [reply]
thanks!
